src/compact.ts

```diff
@@ -2,16 +2,29 @@ function isObject(val) {
   return Object === val.constructor;
 }
 
-export function compact($, keysThatMustExist = [], dropValues: any[] = [undefined], keep = false) {
+export function compact(
+  $,
+  ifCondition = [],
+  options: { dropValues?: any[], $?: any, $1?: any } = {},
+  keep = false
+) {
+  const { dropValues = [undefined], $: original, $1: auxiliary } = options;
   if (Array.isArray($)) {
     $ = $.filter(o => dropValues.indexOf(o) === -1);
     if ($.length > 0) return $;
   } else if ($ && isObject($)) {
+    const keysThatMustExist = ifCondition.filter(a => typeof a === 'string');
     if (keysThatMustExist?.length > 0) {
       for (let key of keysThatMustExist) {
         if (dropValues.indexOf($[key]) !== -1) return;
       }
     }
+    const callbacks = ifCondition.filter(a => typeof a === 'function');
+    if (callbacks?.length > 0) {
+      for (let callback of callbacks) {
+        if (!callback(original, auxiliary)) return;
+      }
+    }
```

Comment:
I would move this logic [here](https://github.com/nikostoulas/metamorphosi/blob/d1318cc10c56e2e0210f83aabe83b47c8242a94c/src/transform.ts#L120) using transform as in callbacks [here](https://github.com/nikostoulas/metamorphosi/blob/d1318cc10c56e2e0210f83aabe83b47c8242a94c/src/transform.ts#L60) because transform handles more cases and by name is the place I would expect to "transform" user's input. I would also try catch it there to avoid a callback discarding the whole transformation.

This way you wouldn't need to pass original, auxiliary here (changing compact's signature). You would have to look for booleans and falsy values here only.

However, this is internal to the lib and could be refactored in the future.

Reply:
Indeed, that makes sense. However, my main motivation for my initial implementation was to have the behavior of $if written in a single point (thus I chose to have it inside the compact.ts). If I understood correctly, by checking my latest commit you will notice now that I split the contents of $if by passing the strings to the compact.ts (thus having the compact.ts completely unaffected), and the callbacks contained are handled [here](https://github.com/nikostoulas/metamorphosi/blob/d1318cc10c56e2e0210f83aabe83b47c8242a94c/src/transform.ts#L120) as you suggested.